Readme.md
# Cluster

 [Cluster](http://learnboost.github.com/cluster) is an extensible multi-core server manager for [node.js](http://nodejs.org).

## Installation

```bash
$ npm install cluster
```

## Features

  - zero-downtime restart
  - hard shutdown support
  - graceful shutdown support
  - resuscitates workers
  - maintains worker count, even if worker was _SIGKILL_ed.
  - workers commit suicide when master dies 
  - spawns one worker per cpu (by default)
  - extensible via plugins
  - bundled plugins
    - [cli](http://learnboost.github.com/cluster/docs/cli.html): provides a command-line interface for your cluster
    - [debug](http://learnboost.github.com/cluster/docs/debug.html): verbose debugging information
    - [logger](http://learnboost.github.com/cluster/docs/logger.html): master / worker logs
    - [pidfiles](http://learnboost.github.com/cluster/docs/pidfiles.html): writes master / worker pidfiles
    - [reload](http://learnboost.github.com/cluster/docs/reload.html): reloads workers when files change
    - [repl](http://learnboost.github.com/cluster/docs/repl.html): perform real-time administration
    - [stats](http://learnboost.github.com/cluster/docs/stats.html): adds real-time statistics to the `repl` plugin
  - supports node 0.2.x
  - supports node 0.4.x
  - supports TCP servers

## Example

app.js:

```javascript
var http = require('http');

module.exports = http.createServer(function(req, res){
  console.log('%s %s', req.method, req.url);
  var body = 'Hello World';
  res.writeHead(200, { 'Content-Length': body.length });
  res.end(body);
});
```

server.js:

```javascript
var cluster = require('cluster')
  , app = require('./app');

cluster(app)
  .use(cluster.logger('logs'))
  .use(cluster.stats())
  .use(cluster.pidfiles('pids'))
  .use(cluster.cli())
  .use(cluster.repl(8888))
  .listen(3000);
```

Note that cluster does _not_ create these directories for you, so you may want to:

    $ mkdir {logs,pids}

recommended usage: passing the path to prevent unnecessary database connections in the master process, as `./app` is only `require()`ed within the workers.

```javascript
var cluster = require('cluster');

cluster('./app')
  .use(cluster.logger('logs'))
  .use(cluster.stats())
  .use(cluster.pidfiles('pids'))
  .use(cluster.cli())
  .use(cluster.repl(8888))
  .listen(3000);
```

## Plugins

 Below are the known 3rd-party plugins for cluster:
 
   - [cluster-log](https://github.com/LearnBoost/cluster-log) remote logger powered by redis
   - [cluster-mail](https://github.com/LearnBoost/cluster-mail) email exception notifications
   - [cluster-exception](https://github.com/3rd-eden/cluster.exception) extensive exception notifications
   - [cluster-responsetimes](https://github.com/mnutt/cluster-responsetimes) response time statistics for cluster's REPL

## Screencasts

  - Cluster [Introduction](http://screenr.com/X8v)

## Running Tests

Install development dependencies:

```bash
$ npm install
```

Then:

```bash
$ make test
```

Actively tested with node:

  - 0.4.9

## Authors

  * TJ Holowaychuk

## License 

(The MIT License)

Copyright (c) 2011 LearnBoost &lt;dev@learnboost.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

0.7.7 / 2011-10-04 
==================

  * Fixed `repl()` bug: ensure server exists before closing it

0.7.6 / 2011-10-03 
==================

  * Fixed standalone read from stdin (__ENOTSOCK__) regression. Closes #153

0.7.5 / 2011-09-23 
==================

  * Fixed an `err.code` check

0.7.4 / 2011-09-19 
==================

  * Fixed `.listen()` return value. Closes #149 [nibblebot]
  * Fixed `cli()` hang regression. Closes #148

0.7.3 / 2011-09-12 
==================

  * Changed: nextTick() uncaughtException handler

0.7.2 / 2011-09-12 
==================

  * Changed `reload()` to reload workers only
  * Changed: use preventDefault instead of `exit()` in `cli()`
  * Fixed "listening" event by deferring "start"
  * Removed local socket usage as it is being removed from node

0.7.1 / 2011-08-18 
==================

  * Added vhost example. Closes #144
  * Fixed `cli()` plugin. Closes #145 [reported by alefnula]

0.7.0 / 2011-08-15 
==================

  * Fixed stand-alone support with a file path. Closes #141 [reported by SebastianEdwards]
  * Fixed clobbering of "worker exception" [reported by fredericosilva]
  * Fixed `logger()` / `pidfiles()` errors when dir does not exist, now using
  * Removed remaining 2.x support. Closes #108
mkdirp. Closes #783

0.6.9 / 2011-07-20 
==================

  * Fixed typo in `reload()` plugin causing the `signal` option to fail. Closes #131

0.6.8 / 2011-07-19 
==================

  * Removed unnecessary use of client socket causing `bind()` errors

0.6.7 / 2011-07-19 
==================

  * Added test case for #125 [felixge]
  * Expose child `Worker` instance `.worker`
  * Fixed restart race-condition. Closes #125

0.6.6 / 2011-07-18 
==================

  * Added stand-alone restart test
  * Added stand-alone shutdown test
  * Added simple stand-alone test
  * Removed 0.2.x compatibility
  * Changed: sync unlink of server / client sockets
  * Fixed stand-alone issue with not killing the parent master. Closes #565

0.6.5 / 2011-07-18 
==================

  * Added support for calling any master method from a worker [felixge]
  * Changed internal IPC to use UDP. Closes #126
  * Fixed: __SIGKILL__ children on master uncaught exception

0.6.4 / 2011-06-14 
==================

  * Fix for json framing. Closes #109

0.6.3 / 2011-06-11 
==================

  * Added `{ color: false }` option to `debug()`
  * Fixed; close socketpair fds when worker dies
  * Fixed `Master#listen()` with env specific config. Closes #98

0.6.2 / 2011-05-11 
==================

  * Fixed IPC for workers without a server. Closes #91
  * Fixed `close(fd)` issue for Master without a server. Closes #89

0.6.1 / 2011-04-26 
==================

  * Changed; cli commands will now signal orphaned children
  * Changed; postpone spawning until "listening" this _should_ fix our EINVAL issue
  * Changed; exit > 0 when trying to use the `cli()` when cluster is not running
  * Changed; `cli()` will still operate on orphans

0.6.0 / 2011-04-18 
==================

  * Added support to run cluster without a server. Closes #72
  * Renamed titles to "cluster" and "cluster worker". closes #82

0.5.7 / 2011-04-17 
==================

  * Added `lightRequests` option to `stats()`

0.5.6 / 2011-04-15 
==================

  * Added; expose utils, helpful for plugins
  * Added; default both `Master#spawn()` and `Master#remove()` to 1

0.5.5 / 2011-04-05 
==================

  * Revert "Changed; demote user/group in master"

0.5.4 / 2011-04-05 
==================

  * Added `title` and `worker title` settings. Closes #54
  * Added `request complete` `stats()` event
  * Changed; demote user/group in master

0.5.3 / 2011-03-30 
==================

  * Added support for changing watched file extensions [Eiríkur Nilsson]
  * Fixed; reload() using extname() instead of indexOf() [reported by solsys]

0.5.1 / 2011-03-24 
==================

  * Changed; only caught uncaughtExceptions when no other listeners are present

0.5.0 / 2011-03-24 
==================

  * Added `connections` option to `stats()` plugin.
    Reports connections and disconnections, displaying in the REPL.
  * Added `requests` option to `stats()` plugin.
    Reports request statistics, displaying in the REPL.
  * Added support for plugins to work within workers. Closes #27
  * Fixed json framing race-condition. Closes #64

0.4.2 / 2011-03-15 
==================

  * Fixed `user` / `group` options. Closes #60
  * Fixed; abort on many immediate worker deaths within boot
  * Fixed `cli()` exit when working with `reload()` (or anything else keeping the event loop active)

0.4.1 / 2011-03-10 
==================

  * Added cyclic restart timeouts. Closes #23
  * Remove master __SIGHUP__ as restart

0.4.0 / 2011-03-08 
==================

  * Added `worker removed` event
  * Added `spawn(-n, signal)` support defaulting to __SIGQUIT__
  * Added `spawn(-n)` support. Closes #46

0.3.3 / 2011-03-03 
==================

  * Added __CLUSTER_WORKER___ env var with the workers id

0.3.2 / 2011-03-01 
==================

  * Fixed bug when using `cluster(filename)`, previously still requiring for master

0.3.1 / 2011-02-28 
==================

  * Added `cluster(filename)` support. Closes #45
    This is highly recommended, view the API docs
    on the site for more info.

0.3.0 / 2011-02-28 
==================

  * Added "worker exception" event. Closes #41
  * Added `listen()` host dns resolution. Closes #35

lib/master.js

/*!
 * Cluster - Master
 * Copyright(c) 2011 LearnBoost <dev@learnboost.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var Worker = require('./worker')
  , EventEmitter = require('events').EventEmitter
  , dirname = require('path').dirname
  , spawn = require('child_process').spawn
  , utils = require('./utils')
  , fsBinding = process.binding('fs')
  , netBinding = process.binding('net')
  , bind = netBinding.bind
  , listen = netBinding.listen
  , socket = netBinding.socket
  , socketpair = netBinding.socketpair
  , close = netBinding.close
  , unlink = fsBinding.unlink
  , dgram = require('dgram')
  , tty = require('tty')
  , net = require('net')
  , fs = require('fs')
  , os = require('os');

/**
 * Node binary.
 */

var node = process.execPath;

/**
 * Start a new `Master` with the given `server` or filename to
 * a node module exporting a server.
 *
 * Options:
 *
 *   - `workers`  Number of workers to spawn, defaults to the number of CPUs
 *   - 'working directory`  Working directory defaulting to the script's dir
 *   - 'backlog` Connection backlog, defaulting to 128
 *   - 'socket port` Master socket port defaulting to `8989`
 *   - 'timeout` Worker shutdown timeout in milliseconds, defaulting to 60,000
 *   - 'user` User id / name
 *   - 'group` Group id / name
 *   - `title` Master process title, defaults to "cluster master"
 *   - `worker title` Worker process title, defaults to "cluster worker {n}"
 *
 * Events:
 *
 *   - `start`. When the IPC server is prepped
 *   - `worker`. When a worker is spawned, passing the `worker`
 *   - `listening`. When the server is listening for connections  
 *   - `closing`. When master is shutting down
 *   - `close`. When master has completed shutting down
 *   - `worker killed`. When a worker has died
 *   - `worker exception`. Worker uncaughtException. Receives the worker / exception
 *   - `worker removed`. Worker removed via `spawn(-n)`
 *   - `kill`. When a `signal` is being sent to all workers
 *   - `restarting`. Restart requested by REPL or signal. Receives an object
 *      which can be patched in order to preserve plugin state.
 *   - `restart`. Restart complete, new master established, previous died.
 *      Receives an object with state preserved by the `restarting` event.
 *                
 * Signals:
 *
 *   - `SIGINT`   hard shutdown
 *   - `SIGTERM`  hard shutdown
 *   - `SIGQUIT`  graceful shutdown
 *   - `SIGUSR2`  graceful restart
 *
 * @param {net.Server|String} server
 * @return {Master}
 * @api public
 */

var Master = module.exports = function Master(server) {
  var self = this;
  this.server = server;
  this.plugins = [];
  this.children = [];
  this.state = 'active';
  this.startup = new Date;
  this._killed = 0;

  // grab server root
  this.cmd = process.argv.slice(1);
  this.dir = dirname(this.cmd[0]);

  // environment
  this.env = process.env.NODE_ENV || 'development';

  // defaults
  this.options = {
      'backlog': 128
    , 'working directory': this.dir
    , 'socket port': 8989
    , 'socket addr': '127.0.0.1'
    , 'timeout': 60000
    , 'restart threshold': 'development' == this.env ? 5000 : 60000
    , 'restart timeout': 'development' == this.env ? 5000 : 60000
    , 'title': 'cluster'
    , 'worker title': 'cluster worker'
  };

  // parent master pid
  this.ppid = process.env.CLUSTER_PARENT_PID
    ? parseInt(process.env.CLUSTER_PARENT_PID, 10)
    : null;

  // process is a worker
  this.isWorker = !! process.env.CLUSTER_MASTER_PID;

  // process is a child (worker or master replacement)
  this.isChild = this.isWorker || !! process.env.CLUSTER_REPLACEMENT_MASTER;

  // process is master
  this.isMaster = ! this.isWorker;

  // process id
  this.pid = process.pid;
  if (this.isMaster) process.env.CLUSTER_MASTER_PID = this.pid;

  // custom worker fds, defaults to std{out,err}
  this.customFds = [1, 2];

  // resolve server filename
  if (this.isWorker && 'string' == typeof this.server) {
    this.server = require(this.resolve(this.server));
  }

  // IPC is prepped
  this.on('start', function(){
    process.chdir(self.options['working directory']);
  });

  // spawn our workers
  this.on('listening', function(){
    self.spawn(self.options.workers);
    self.listening = true;
  });

  // kill children on master exception
  if (this.isMaster) {
    process.on('uncaughtException', function(err){
      self.kill('SIGKILL');
      console.error(err.stack || String(err));
      process.exit(1);
    });
  }
};

/**
 * Interit from `EventEmitter.prototype`.
 */

Master.prototype.__proto__ = EventEmitter.prototype;

/**
 * Worker is a receiver.
 */

require('./mixins/receiver')(Master.prototype);

/**
 * Resolve `path` relative to the server file being executed.
 *
 * @param {String} path
 * @return {String}
 * @api public
 */

Master.prototype.resolve = function(path){
  return '/' == path[0]
    ? path
    : this.dir + '/' + path;
};

/**
 * Return `true` when the environment set by `Master#in()`
 * matches __NODE_ENV__.
 *
 * @return {Boolean}
 * @api private
 */

Master.prototype.__defineGetter__('environmentMatches', function(){
  if (this._env) 
    return this.env == this._env || 'all' == this._env;
  return true;
});

/**
 * Invoke masters's `method` with worker `id`. (called from Worker)
 *
 * @param {Number} id
 * @param {String} method
 * @param {...} args
 * @api private
 */

Master.prototype.call = function(id, method){
  this.sock = this.sock || dgram.createSocket('udp4');

  var msg = new Buffer(utils.frame({
      args: utils.toArray(arguments, 2)
    , method: method
    , id: id
  }));

  this.sock.send(
      msg
    , 0
    , msg.length
    , this.options['socket port']
    , this.options['socket addr']);
};

/**
 * Perform setup tasks then invoke `fn()` when present.
 *
 * @param {Function} fn
 * @return {Master} for chaining
 * @api public
 */

Master.prototype.start = function(fn){
  var self = this;

  // deferred title
  process.title = this.options.title;

  // prevent listen
  if (this.preventDefault) return this;

  // env match
  if (this.environmentMatches) {
    // worker process
    if (this.isWorker) {
      this.worker = new Worker(this);
      this.worker.start();
    // master process
    } else if (fn) {
      fn();
    // standalone
    } else {
      this.on('start', function(){ self.emit('listening'); });
      if (this.isChild) this.acceptFd();
      this.setupIPC();
    }
  }

  return this;
};

/**
 * Defer `http.Server#listen()` call.
 *
 * @param {Number|String} port or unix domain socket path
 * @param {String|Function} host or callback
 * @param {Function} callback
 * @return {Master} for chaining
 * @api public
 */

Master.prototype.listen = function(port, host, callback){
  var self = this;
  if (!this.environmentMatches) return this;
  if ('function' == typeof host) callback = host, host = null;
  this.port = port;
  this.host = host;
  this.callback = callback;
  return this.start(function(){
    self.on('start', function(){
      self.startListening(!self.isChild);
    });

    if (self.isChild) {
      self.acceptFd();
    } else {
      self.createSocket(function(err, fd){
        if (err) throw err;
        self.fd = fd;
        self.setupIPC();
      });
    }
  });
};

/**
 * Create / return IPC socket.
 *
 * @api private
 */

Master.prototype.IPCSocket = function(){
  var self = this;
  if (this._sock) return this._sock;
  this._sock = dgram.createSocket('udp4');
  this._sock.on('message', function(msg, info){
    try {
      msg = JSON.parse(msg.toString('ascii'));
      self.invoke(msg.method, msg.args, self.children[msg.id]);
    } catch (err) {
      console.error(err.stack || String(err));
    }
  });
  return this._sock;
};

/**
 * Setup IPC.
 *
 * @api private
 */

Master.prototype.setupIPC = function(){
  var self = this;

  // signal handlers
  this.registerSignalHandlers();

  // Default worker to the # of cpus
  this.defaultWorkers();

  // udp server for IPC
  this.IPCSocket().on('listening', function(){
    process.nextTick(function(){
      self.emit('start');
    });
  });

  // bind
  this.IPCSocket().bind(
      this.options['socket port']
    , this.options['socket addr']);
};

/**
 * Conditionally perform the following action, if 
 * __NODE_ENV__ matches `env`.
 *
 * Examples:
 *
 *      cluster(server)
 *        .in('development').use(cluster.debug())
 *        .in('development').listen(3000)
 *        .in('production').listen(80);
 *     
 * @param {String} env
 * @return {Master} self or stubs
 * @api public
 */

Master.prototype.in = function(env){
  this._env = env;
  return this;
};

/**
 * Set option `key` to `val`.
 *
 * @param {String} key
 * @param {Mixed} val
 * @return {Master} for chaining
 * @api public
 */

Master.prototype.set = function(key, val){
  if (this.environmentMatches) this.options[key] = val;
  return this;
};

/**
 * Invoke `fn(master)`.
 *
 * @param {Function} fn
 * @api public
 */

Master.prototype.do = function(fn){
  if (this.environmentMatches) fn.call(this, this);
  return this;
};

/**
 * Check if `option` has been set.
 *
 * @param {String} option
 * @return {Boolean}
 * @api public
 */

Master.prototype.has = function(option){
  return !! this.options[option];
};

/**
 * Use the given `plugin`.
 *
 * @param {Function} plugin
 * @return {Master} for chaining
 * @api public
 */

Master.prototype.use = function(plugin){
  if (this.environmentMatches) {
    this.plugins.push(plugin);
    if (this.isWorker) {
      plugin.enableInWorker && plugin(this);
    } else {
      plugin(this);
    }
  }
  return this;
};

/**
 * Create listening socket and callback `fn(err, fd)`.
 *
 * @return {Function} fn
 * @api private
 */

Master.prototype.createSocket = function(fn){
  var self = this
    , ipv;

  // explicit host
  if (this.host) {
    // ip
    if (ipv = net.isIP(this.host)) {
      fn(null, socket('tcp' + ipv));
    // lookup
    } else {
      require('dns').lookup(this.host, function(err, ip, ipv){
        if (err) return fn(err);
        self.host = ip;
        fn(null, socket('tcp' + ipv));
      });
    }
  // local socket
  } else if ('string' == typeof this.port) {
    fn(null, socket('unix'));
  // only port
  } else if ('number' == typeof this.port) {
    fn(null, socket('tcp4'));
  }
};

/**
 * Register signal handlers.
 *
 * @api private
 */

Master.prototype.registerSignalHandlers = function(){
  var self = this;
  process.on('SIGINT', this.destroy.bind(this));
  process.on('SIGTERM', this.destroy.bind(this));
  process.on('SIGQUIT', this.close.bind(this));
  process.on('SIGUSR2', this.attemptRestart.bind(this));
  process.on('SIGCHLD', this.maintainWorkerCount.bind(this));
};

/**
 * Default workers to the number of cpus available.
 *
 * @api private
 */

Master.prototype.defaultWorkers = function(){
  if (!this.has('workers')) {
    this.set('workers', os
      ? os.cpus().length
      : 1);
  }
};

/**
 * Restart workers only, sending `signal` defaulting
 * to __SIGQUIT__.
 *
 * @param {Type} name
 * @return {Type}
 * @api public
 */

Master.prototype.restartWorkers = function(signal){
  this.kill(signal || 'SIGQUIT');
};

/**
 * Maintain worker count, re-spawning if necessary.
 *
 * @api private
 */

Master.prototype.maintainWorkerCount = function(){
  this.children.forEach(function(worker){
    var pid = worker.proc.pid;
    if (!pid) this.workerKilled(worker);
  }, this);
};

/**
 * Remove `n` workers with `signal`
 * defaulting to __SIGQUIT__.
 *
 * @param {Number} n
 * @param {String} signal
 * @api public
 */

Master.prototype.remove = function(n, signal){
  if (!arguments.length) n = 1;
  var len = this.children.length
    , worker;

  // cap at worker len
  if (n > len) n = len;

  // remove the workers
  while (n--) {
    worker = this.children.pop();
    worker.proc.kill(signal || 'SIGQUIT');
    this.emit('worker removed', worker);
    this.removeWorker(worker.id);
  }
};

/**
 * Remove worker `id`.
 *
 * @param {Number} id
 * @api public
 */

Master.prototype.removeWorker = function(id){
  var worker = this.children[id];
  if (!worker) return;
  if (worker.fds) {
    close(worker.fds[0]);
    close(worker.fds[1]);
  }
  delete this.children[id];
};

/**
 * Spawn `n` workers.
 *
 * @param {Number} n
 * @api public
 */

Master.prototype.spawn = function(n){
  if (!arguments.length) n = 1;
  while (n--) this.spawnWorker();
};

/**
 * Spawn a worker with optional `id`.
 *
 * @param {Number} id
 * @return {Worker}
 * @api private
 */

Master.prototype.spawnWorker = function(id){
  var worker;

  // id given
  if ('number' == typeof id) {
    worker = new Worker(this).spawn(id)
    this.children[id] = worker;
    worker.id = id;
  // generate an id
  } else {
    worker = new Worker(this).spawn(this.children.length);
    this.children.push(worker);
  }

  var obj = {
      method: 'connect'
    , args: [worker.id, this.options]
  };

  worker.sock.write(utils.frame(obj), 'ascii', this.fd);

  // emit
  this.emit('worker', worker);

  return worker;
};

/**
 * Graceful shutdown, wait for all workers
 * to reply before exiting.
 *
 * @api public
 */

Master.prototype.close = function(){
  this.state = 'graceful shutdown';
  this.emit('closing');
  this.kill('SIGQUIT');
  this.pendingDeaths = this.children.length;
};

/**
 * Hard shutdwn, immediately kill all workers.
 *
 * @api public
 */

Master.prototype.destroy = function(){
  this.state = 'hard shutdown';
  this.emit('closing');
  this.kill('SIGKILL');
  this._destroy();
};

/**
 * Attempt restart, while respecting the `restart threshold`
 * setting, to help prevent recursive restarts.
 *
 * @param {String} sig
 * @api private
 */

Master.prototype.attemptRestart = function(sig){
  var uptime = new Date - this.startup
    , threshold = this.options['restart threshold']
    , timeout = this.options['restart timeout'];

  if (this.__restarting) return;

  if (uptime < threshold) {
    this.__restarting = true;
    this.emit('cyclic restart');
    setTimeout(function(self){
      self.restart(sig);
    }, timeout, this);
  } else {
    this.restart(sig);
  }
};

/**
 * Restart all workers, by sending __SIGQUIT__
 * or `sig` to them, enabling master to re-spawn.
 *
 * @param {String} sig
 * @return {ChildProcess} replacement master process
 * @api public
 */

Master.prototype.restart = function(sig){
  var data = {}
    , proc = this.spawnMaster();

  // pass object to plugins, allowing them
  // to patch it, and utilize the data in
  // the new Master
  this.emit('restarting', data);
  proc.sock.write(utils.frame({
      method: 'connectMaster'
    , args: [sig || 'SIGQUIT']
  }), 'ascii', this.fd);

  this.on('close', function(){
    proc.sock.write(utils.frame({
        method: 'masterKilled'
      , args: [data]
    }), 'ascii');
  });

  return proc;
};

/**
 * Spawn a new master process.
 *
 * @return {ChildProcess}
 * @api private
 */

Master.prototype.spawnMaster = function(){
  var fds = socketpair()
    , customFds = [fds[0], 1, 2]
    , env = {};

  // merge current env
  for (var key in process.env) {
    env[key] = process.env[key];
  }

  delete env.CLUSTER_MASTER_PID;
  env.CLUSTER_REPLACEMENT_MASTER = 1;
  env.CLUSTER_PARENT_PID = this.pid;

  // spawn new master process
  var proc = spawn(node, this.cmd, {
      customFds: customFds
    , env: env
  });
  
  // unix domain socket for ICP + fd passing
  proc.sock = new net.Socket(fds[1], 'unix');
  return proc;
};

/**
 * Master replacement connected.
 *
 * @param {String} sig
 * @api private
 */

Master.prototype.connectMaster = function(sig){
  var self = this;

  function kill(){
    process.kill(self.ppid, sig);
  }

  if (this.listening) return kill();
  this.on('listening', kill);
};

/**
 * Original master has died aka 'retired',
 * we now fire the 'restart' event.
 *
 * @param {Object} data
 * @api private
 */

Master.prototype.masterKilled = function(data){
  this.emit('restart', data);
};

/**
 * Accept fd from parent master, then `setupIPC()`.
 *
 * @api private
 */

Master.prototype.acceptFd = function(){
  var self = this
    , stdin = new net.Socket(0, 'unix');

  // set fd and start master
  stdin.setEncoding('ascii');
  stdin.on('fd', function(fd){
    self.fd = fd;
    self.setupIPC();
  });

  // frame commands from the parent master
  stdin.on('data', this.frame.bind(this));
  stdin.resume();
};

/**
 * Close servers and emit 'close' before exiting.
 *
 * @api private
 */

Master.prototype._destroy = function(){
  this.IPCSocket().close();
  if (this.fd) close(this.fd);
  this.emit('close');
  process.nextTick(process.exit.bind(process));
};

/**
 * Worker is connected.
 *
 * @param {Worker} worker
 * @api private
 */

Master.prototype.connect = function(worker){
  this.emit('worker connected', worker);
};

/**
 * Start listening, when `shouldBind` is `true` the socket
 * will be bound, and will start listening for connections.
 *
 * @param {Boolean} shouldBind
 * @api private
 */

Master.prototype.startListening = function(shouldBind){
  var self = this;

  // remove unix domain socket 
  if ('string' == typeof this.port && shouldBind) {
    fs.unlink(this.port, function(err){
      if (err && 'ENOENT' != err.code) throw err;
      startListening();
    });
  } else {
    startListening();
  }

  // bind / listen
  function startListening() {
    if (shouldBind) {
      try {
        bind(self.fd, self.port, self.host);
        listen(self.fd, self.options.backlog);
      } catch(e) {
        self.kill('SIGKILL');
        throw e;
      }
    }
    self.callback && self.callback();
    self.emit('listening');
  }
};

/**
 * The given `worker` has been killed.
 * Emit the "worker killed" event, remove
 * the worker, and re-spawn depending on 
 * the master state.
 *
 * @api private
 */

Master.prototype.workerKilled = function(worker){
  // if we have many failing workers at boot
  // then we likely have a serious issue.
  if (new Date - this.startup < 20000) {
    if (++this._killed == 20) {
      console.error('');
      console.error('Cluster detected over 20 worker deaths in the first');
      console.error('20 seconds of life, there is most likely');
      console.error('a serious issue with your server.');
      console.error('');
      console.error('aborting.');
      console.error('');
      process.exit(1);
    }
  }

  // emit event
  this.emit('worker killed', worker);

  // always remove worker
  this.removeWorker(worker.id);

  // state specifics
  switch (this.state) {
    case 'hard shutdown':
      break;
    case 'graceful shutdown':
      --this.pendingDeaths || this._destroy();
      break;
    default:
      this.spawnWorker(worker.id);
  }
};

/**
 * `worker` received exception `err`.
 *
 * @api private
 */

Master.prototype.workerException = function(worker, err){
  this.emit('worker exception', worker, err);
};

/**
 * Received worker timeout.
 * 
 * @api private
 */

Master.prototype.workerTimeout = function(worker, timeout){
  this.emit('worker timeout', worker, timeout);
};

/**
 * Worker waiting on `connections` to close.
 * 
 * @api private
 */

Master.prototype.workerWaiting = function(worker, connections){
  this.emit('worker waiting', worker, connections);
};

/**
 * Send `sig` to all worker processes, defaults to __SIGTERM__.
 *
 * @param {String} sig
 * @api public
 */

Master.prototype.kill = function(sig){
  var self = this;
  this.emit('kill', sig);
  this.children.forEach(function(worker){
    worker.proc.kill(sig);
  });
};
